internal/searchindex/searchindex.go

```diff
@@ -0,0 +1,440 @@
+	//
+	//d := cmp2.Diff(copySelf, copyValue, cmpopts.EquateEmpty())
+	//fmt.Println("INDEX DIFF", d)
+
+	return cmp.SemanticEqual(copySelf, copyValue)
+}
+
+func (s *SearchIndex) Normalize() *SearchIndex {
+	// TODO: Refactor interface to return error!
+	_ = cmp.Normalize(s)
+	return s
+}
+
+func (s *SearchIndex) ToAtlas() (*admin.ClusterSearchIndex, error) {
+	convertJSONToListOfMaps := func(in *apiextensionsv1.JSON) (*[]map[string]interface{}, error) {
+		if in == nil {
+			return nil, nil
+		}
+		var result []map[string]interface{}
+		err := json.Unmarshal(in.Raw, &result)
+		return &result, err
+	}
+
+	convertJSONToMap := func(in *apiextensionsv1.JSON) (map[string]interface{}, error) {
+		if in == nil {
+			return nil, nil
+		}
+		result := map[string]interface{}{}
+		err := json.Unmarshal(in.Raw, &result)
+		return result, err
+	}
+	convertJSONToInterface := func(in *apiextensionsv1.JSON) (*[]interface{}, error) {
+		if in == nil {
+			return pointer.MakePtr([]interface{}{}), nil
+		}
+		var result []interface{}
+		err := json.Unmarshal(in.Raw, &result)
+		return &result, err
+	}
+
+	storedSource, err := convertJSONToMap(s.StoredSource)
+	if err != nil {
+		return nil, fmt.Errorf("unable to convert storedSource: %w", err)
+	}
+
+	analyzers, err := func(in *[]akov2.AtlasSearchIndexAnalyzer) (*[]admin.ApiAtlasFTSAnalyzers, error) {
+		if in == nil {
+			return nil, nil
+		}
+
+		result := make([]admin.ApiAtlasFTSAnalyzers, 0, len(*in))
+		for i := range *in {
+			analyzer := (*in)[i]
+			charFilters, err := convertJSONToInterface(analyzer.CharFilters)
+			if err != nil {
+				return nil, err
+			}
+
+			tokenFilters, err := convertJSONToInterface(analyzer.TokenFilters)
+			if err != nil {
+				return nil, err
+			}
+
+			result = append(result, admin.ApiAtlasFTSAnalyzers{
+				CharFilters:  charFilters,
+				Name:         analyzer.Name,
+				TokenFilters: tokenFilters,
+				Tokenizer: admin.ApiAtlasFTSAnalyzersTokenizer{
+					MaxGram:        analyzer.Tokenizer.MaxGram,
+					MinGram:        analyzer.Tokenizer.MinGram,
+					Type:           analyzer.Tokenizer.Type,
+					Group:          analyzer.Tokenizer.Group,
+					Pattern:        analyzer.Tokenizer.Pattern,
+					MaxTokenLength: analyzer.Tokenizer.MaxTokenLength,
+				},
+			})
+		}
+		return &result, nil
+	}(s.Analyzers)
+	if err != nil {
+		return nil, err
+	}
+
+	mappings, err := func(in *akov2.Mappings) (*admin.ApiAtlasFTSMappings, error) {
+		if in == nil {
+			return nil, nil
+		}
+		fields, err := convertJSONToMap(in.Fields)
+		if err != nil {
+			return nil, err
+		}
+		return &admin.ApiAtlasFTSMappings{
+			Dynamic: in.Dynamic,
+			Fields:  fields,
+		}, nil
+	}(s.Search.Mappings)
+	if err != nil {
+		return nil, err
+	}
+
+	synonyms := func(in *[]akov2.Synonym) *[]admin.SearchSynonymMappingDefinition {
+		if in == nil {
+			return nil
+		}
+
+		result := make([]admin.SearchSynonymMappingDefinition, 0, len(*in))
+		for i := range *in {
+			syn := &(*in)[i]
```

Comment:
please godoc why this is needed

Reply:
is it actually?

Reply:
Not sure what are you asking about. The type in Atlas and the type in AKO are different, that's why there is a convert method

Reply:
Nevermind, I was assuming we're shallow-copying by `&(*in)`. all good, resolving.